app/src/main/java/com/sunasterisk/boringweather/ui/detail/DetailWeatherAdapter.kt

```diff
@@ -0,0 +1,94 @@
+package com.sunasterisk.boringweather.ui.detail
+
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.view.animation.Animation
+import android.view.animation.RotateAnimation
+import androidx.recyclerview.widget.ListAdapter
+import com.sunasterisk.boringweather.base.BaseViewHolder
+import com.sunasterisk.boringweather.ui.detail.model.DailyWeatherItem
+import com.sunasterisk.boringweather.ui.detail.model.DetailWeatherAdapterItem
+import com.sunasterisk.boringweather.ui.detail.model.HourlyWeatherItem
+import kotlinx.android.synthetic.main.item_weather_hourly_expandable.view.*
+
+class DetailWeatherAdapter(
+    private val clickListener: (Int) -> Unit
+) : ListAdapter<DetailWeatherAdapterItem<*>, BaseViewHolder<*>>(
+    DetailWeatherAdapterItem.diffUtil
+) {
+    override fun getItemViewType(position: Int): Int {
+        return getItem(position).viewType.layoutRes
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<*> {
+        val inflater = LayoutInflater.from(parent.context)
+        val view = inflater.inflate(viewType, parent, false)
+        return when (viewType) {
+            DetailWeatherAdapterViewType.HOURLY_WEATHER.layoutRes -> {
+                HourlyWeatherVH(view) { position: Int, expanded: Boolean ->
+                    if (expanded) clickListener(position)
+                    notifyItemChanged(position)
+                }
+            }
+            DetailWeatherAdapterViewType.DAILY_WEATHER.layoutRes -> {
+                DailyWeatherVH(view)
+            }
+            else -> throw IllegalArgumentException("Wrong view type: $viewType")
+        }
+    }
+
+    override fun onBindViewHolder(holder: BaseViewHolder<*>, position: Int) {
+        val item: DetailWeatherAdapterItem<*> = getItem(position)
+        if (holder is HourlyWeatherVH && item is HourlyWeatherItem) {
+            holder.bind(item)
+        } else if (holder is DailyWeatherVH && item is DailyWeatherItem) {
+            holder.bind(item)
+        }
+    }
+
+    class DailyWeatherVH(view: View) : BaseViewHolder<DailyWeatherItem>(view) {
+        override fun bind(item: DailyWeatherItem) {
+            // TODO bind data here
+        }
+    }
+
+    class HourlyWeatherVH(view: View, private val clickListener: (Int, Boolean) -> Unit) :
+        BaseViewHolder<HourlyWeatherItem>(view) {
+        override fun bind(item: HourlyWeatherItem) {
+            itemView.textTime.text = item.data.dateTime.toString() // TODO bind data here
+            itemView.containerDetail.visibility = if (item.expanded) View.VISIBLE else View.GONE
+            itemView.buttonExpand.setOnClickListener {
+                item.expanded = !item.expanded
+                clickListener(adapterPosition, item.expanded)
+                animationPos = adapterPosition
+            }
+            rotateAnimateExpandButton(item)
+        }
```

Comment:
Dùng with tránh lặp code + isVisible 
```suggestion
        override fun bind(item: HourlyWeatherItem) {
            with(itemView){
                textTime.text = item.data.dateTime.toString() // TODO bind data here
                containerDetail.isVisible = item.expanded
                buttonExpand.setOnClickListener {
                    item.expanded = !item.expanded
                    clickListener(adapterPosition, item.expanded)
                    animationPos = adapterPosition
                }            
            }
            
            rotateAnimateExpandButton(item)
        }
```